ai2_internal/api.py

```diff
@@ -34,60 +36,106 @@ class Span(BaseModel):
     box: Optional[Box]
 
     @classmethod
-    def from_mmda(cls, span: mmda_ann.Span) -> Span:
-        ret = cls(start=span.start, end=span.end)
+    def from_mmda(cls, span: mmda_ann.Span) -> "Span":
+        box = Box.from_mmda(span.box) if span.box is not None else None
+        ret = cls(start=span.start, end=span.end, box=box)
         if span.box is not None:
             ret.box = Box.from_mmda(span.box)
         return ret
 
     def to_mmda(self) -> mmda_ann.Span:
-        ret = mmda_ann.Span(start=self.start, end=self.end)
-        if self.box is not None:
-            ret.box = self.box.to_mmda()
-        return ret
+        return mmda_ann.Span.from_json(self.dict())
+
+
+class Attributes(BaseModel):
+    """Class to represent attributes for a SpanGroup or BoxGroup.
+    Attributes are generally stored in mmda as mmda_ann.Metadata objects.
+
+    Unlike mmda.types.annotation.Metadata, this class ignores fields
+    id, type, and text, which are expected to be stored in the SpanGroup
+    / BoxGroup itself."""
+
+    @classmethod
+    def from_mmda(cls, metadata: mmda_ann.Metadata) -> "Attributes":
+        return cls(**metadata.to_json())
+
+    def to_mmda(self) -> mmda_ann.Metadata:
+        return mmda_ann.Metadata.from_json(self.dict())
+
+
+class Annotation(BaseModel, extra=Extra.ignore):
+    attributes: Attributes
+
+    @classmethod
+    def get_metadata_cls(cls) -> Type[Attributes]:
+        for inherit_cls in cls.mro():
+            if not hasattr(inherit_cls, "__annotations__"):
+                continue
+            if "attributes" in inherit_cls.__annotations__:
+                return inherit_cls.__annotations__["attributes"]
```

Comment:
I think we could do this via:

```python
@classmethod
def get_metadata_cls(cls) -> Type[Attributes]:
    return cls.__fields__["attributes"]["type"]
```

rather than using the low level python api and traversing the hierarchy.

Reply:
done!